packages/calypso-e2e/src/secrets/secrets-manager.ts

```diff
@@ -0,0 +1,220 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */ // We have to do some dynamic parsing and iteration as we validate JSON.
+import fs from 'fs';
+import path from 'path';
+
+export const TEST_ACCOUNT_NAMES = [
+	'defaultUser',
+	'eCommerceUser',
+	'simpleSiteFreePlanUser',
+	'simpleSitePersonalPlanUser',
+	'gutenbergSimpleSiteUser',
+	'gutenbergSimpleSiteEdgeUser',
+	'gutenbergAtomicSiteUser',
+	'gutenbergAtomicSiteEdgeUser',
+	'coBlocksSimpleSiteEdgeUser',
+	'siteEditorSimpleSiteUser',
+	'siteEditorSimpleSiteEdgeUser',
+	'martechTosUser',
+	'calypsoPreReleaseUser',
+	'i18nUser',
+	'p2User',
+	'jetpackUser',
+	'jetpackUserPREMIUM',
+	'jetpackUserJN',
+	'commentingUser',
+	'notificationsUser',
+	'googleLoginUser',
+] as const;
+
+export type TestAccountName = typeof TEST_ACCOUNT_NAMES[ number ];
+
+export type OtherTestSiteName = 'notifications';
+
+export interface Secrets {
+	passwordForNewTestSignUps: string;
+	storeSandboxCookieValue: string;
+	testCouponCode: string;
+	wpccAuthPath: string;
+	restApiApplication: {
+		user: string;
+		client_id: string;
+		redirect_uri: string;
+		client_secret: string;
+		access_code: string;
+	};
+	martechTosUploadCredentials: {
+		bearer_token: string;
+	};
+	mailosaur: {
+		apiKey: string;
+		inviteInboxId: string;
+		signupInboxId: string;
+		domainsInboxId: string;
+		defaultUserInboxId: string;
+	};
+	testAccounts: {
+		[ key in TestAccountName ]: {
+			username: string;
+			password: string;
+			primarySite?: string;
+			otherSites?: string[];
+			totpKey?: string;
+		};
+	};
+	otherTestSites: {
+		[ key in OtherTestSiteName ]: string;
+	};
+}
+
+/**
+ * Static class that gives you access to secrets needed for E2E testing.
+ * You must first decrypt the encrypted secrets for tests to be able to access them.
+ * All checking happens at runtime.
+ */
+export class SecretsManager {
+	private static _secretsCache: Secrets;
+
+	/**
+	 * Get the secrets. If this is the first time accessed in the node process, it will initialize and cache the secrets.
+	 */
+	public static get secrets(): Secrets {
+		if ( ! this._secretsCache ) {
+			this._secretsCache = this.initializeSecrets();
+		}
+
+		return this._secretsCache;
+	}
+
+	/**
+	 * Initializes the secrets needed for E2E testing.
+	 */
+	private static initializeSecrets(): Secrets {
+		const SECRETS_FILE_NAME = 'decrypted-secrets.json';
+		const parsedSecrets = this.parseSecretsFile( path.join( __dirname, SECRETS_FILE_NAME ) );
+		const fakeReferenceSecrets = this.createFakeSecretsForValidation();
+		this.validateSecrets( parsedSecrets, fakeReferenceSecrets );
```

Comment:
I __think__ I understand why this is being done.

If someone wants to add a new user for example, then in order to pass the validation they would also need to add the new user to the `fakeSecrets` entry as well, is that correct?

Reply:
>If someone wants to add a new user for example, then in order to pass the validation they would also need to add the new user to the fakeSecrets entry as well, is that correct?

Yes! I was trying to figure out the best way to get...
1. String type safety, and
2. Runtime JSON validation

... and this seemed like the cleanest way to do it for now. 

Basically, if you change the JSON, but not the typings, it will fail validation and runtime.

And if you update the typings, you'll get a transpile error because the fake secrets won't match the typing anymore. So you'll have to update those, and then all your typings and validation will be up to date!

It's the safest approach I could think of, but it's a _little_ clunky IMHO -- any ideas jump out to you of a smoother way to do this?

Reply:
As you say, it isn't the smoothest way but I feel it's a good compromise.

Reply:
The ideal way would be to somehow decrypt, read and get the type structure from the JSON at compile-time, and built the types based on that. This could be fun to explore but wouldn't be trivial (would probably need a TS plugin of sorts), so I think this is a good compromise, too. Great work!